### DISCUSSION TITLE
NPM Docker - How to proxy traffic to Apache installed directly on server

### DISCUSSION THREAD

**Discussion by jeffshead** [NginxProxyManager/nginx-proxy-manager#4072] [Q&A]
My NPM runs in a Docker container and is on the same Docker network as other containers. I also have Apache installed directly on the same host server; non-dockerized. How can I create a NPM proxy host that proxy's traffic to Apache (localhost:82). I'm getting 502's because NPM is on the Docker network.

**Comment by shalafi99** [comment]
For Docker Desktop, you could use host.docker.internal:82

Not entirely sure (I use docker desktop) but with docker engine, it would be "add-host" with the "host-gateway" flag for the NPM docker run command (or for its compose file, it would be the "extra_hosts" element) to achieve a similar result, sources [here](https://docs.docker.com/reference/cli/docker/container/run/#add-host) and [here](https://docs.docker.com/reference/compose-file/build/#extra_hosts)

**Comment by jeffshead** [comment]
I use Portainer. Are you suggesting `host.docker.internal:82` should be the proxy host destination in NPM? That doesn't seem to work on Linux.

**Comment by shalafi99** [comment]
> I use Portainer. Are you suggesting `host.docker.internal:82` should be the proxy host destination in NPM? That doesn't seem to work on Linux.

Yes, but only if Docker Desktop is the docker software managing the containers.

Otherwise, it means the software would be the "regular" docker engine (as it apparently seems to be your case, I use docker desktop in Windows) and for that scenario it would be that additional host parameter provided during container creation.

Are you using a compose file or a docker run command to create the NPM container?

**Comment by jeffshead** [comment]
>Are you using a compose file or a docker run command to create the NPM container?

Portainer handles the containers with it's "stacks" so I'm guessing the answer to your question is Compose.

**Comment by shalafi99** [comment]
Yes, sorry about that - I should have edited my previous reply to include a mention of where in portainer each of those 2 methods is covered by.

I gave it a try myself but had only partial success (_I just created a new proxy host pointing to an existing service, the original proxy host uses host.docker.internal which is available to me on account of Docker Desktop_), here is what I did so far:

Adding this "extra_hosts" to the compose file:

```
  npm-ext:
    image: docker.io/jc21/nginx-proxy-manager:latest
    container_name: npm-ext
    extra_hosts:
      - hosttest.docker.internal:host-gateway
```

Resulted in this additional name resolution (note the 2nd curl) being available from within the NPM container:

![image](https://github.com/user-attachments/assets/20a818b4-224a-4477-8601-b43ae35c1282)

I wrote partial success because despite of curl working as expected inside the container, I am getting a 502 error when I hit the new proxy host using my web browser 🙄. Later I will troubleshoot some more or try to bring some brand new web service up and check again.

**Comment by jeffshead** [comment]
Adding ` - host.docker.internal:host-gateway` to the stack and using `host.docker.internal:82` in the proxy host just doesn't work for me. Actually, when I add...
```
extra_hosts:
      - host.docker.internal:host-gateway
```
...to the compose file (Portainer stack), Portainer fails to start the container.

**Reply by shalafi99** [reply]
There's gotta be some other difference between the docker engine in Linux (yours) and docker desktop for windows (mine) then, I did that test also using portainer. There is some extra info about the value for "host-gateway" [here](https://docs.docker.com/reference/cli/dockerd/#configure-host-gateway-ip), but it seemed applicable only when one wants to change its default behavior.

Did portainer throw any specific error while trying to start the container?
Just to be safe, did you also try with some other "anything.internal" to the left of the colon for the extra_hosts parameter? It would be quite weird for it to fluke exactly because you picked a FQDN which is reserved in docker desktop (which is not what you are using), but who knows.
